compiler/qsc_circuit/src/circuit_to_qsharp.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#[cfg(test)]
+mod tests;
+
+use regex_lite::{Captures, Regex};
+use rustc_hash::FxHashMap;
+
+use crate::{
+    circuit::{CircuitGroup, Measurement, Unitary},
+    Circuit, Operation,
+};
+
+pub fn circuits_to_qsharp(file_name: String, circuits_json: String) -> Result<String, String> {
+    match serde_json::from_str::<CircuitGroup>(circuits_json.as_str()) {
+        Ok(circuits) => Ok(build_circuits(file_name, circuits.circuits)),
+        Err(e) => Err(format!("Error: {}", e)),
+    }
+}
+
+pub fn build_circuits(file_name: String, circuits: Vec<Circuit>) -> String {
+    if circuits.len() == 1 {
+        build_operation_def(file_name, &circuits[0])
+    } else {
+        let mut qsharp_str = String::new();
+        for (index, circuit) in circuits.iter().enumerate() {
+            let circuit_name = format!("{file_name}{index}");
+            let circuit_str = build_operation_def(circuit_name, circuit);
+            qsharp_str.push_str(&circuit_str);
+        }
+        qsharp_str
+    }
+}
+
+pub fn build_operation_def(circuit_name: String, circuit: &Circuit) -> String {
+    let mut indentation_level = 0;
+    let qubits = circuit
+        .qubits
+        .iter()
+        .enumerate()
+        .map(|(i, q)| (q.id, format!("qs[{}]", i)))
+        .collect::<FxHashMap<_, _>>();
+
+    let parameter = if qubits.is_empty() {
+        String::new()
+    } else {
+        "qs : Qubit[]".to_string()
+    };
+
+    // The return type is determined by the number of qubits "children".
+    // However, the actual return statement is determined by the variables storing measurements.
+    // If there is an inconsistency between these, which would happen if there was a mismatch between
+    // the number of qubit children specified on the circuit and the number of qubit children specified
+    // on the measurements, incorrect Q# could be generated.
+    let return_type = match circuit.qubits.iter().fold(0, |sum, q| sum + q.num_results) {
+        0 => "Unit",
+        1 => "Result",
+        _ => "Result[]",
+    };
+
+    // Check if all operations are Unitary
+    let is_ctl_adj = circuit.component_grid.iter().all(|col| {
+        col.components.iter().all(|op| {
+            if let Operation::Unitary(unitary) = op {
+                unitary.gate != "|0〉" && unitary.gate != "|1〉"
+            } else {
+                false
+            }
+        })
```

Comment:
This could be an `.any` that checks for any instances of resets or measurements and stops rather than verifying that every operation isn't reset or measurement.

Reply:
done